analysers/analyser_osmosis_highway_deadend.py

```diff
@@ -71,121 +83,216 @@
+  NOT is_area AND
+  NOT is_construction
+"""
+
+sql31 = """
+CREATE INDEX idx_oneway_highway_linestring ON oneway_highway USING GIST(linestring) WHERE nid_index=1;
+"""
+
+sql32 = """
+-- Collect all nodes that could be possible (valid) end nodes
+CREATE TEMP TABLE allowed_end_nodes AS
+SELECT DISTINCT
+  unnest(nodes) AS id
+FROM
+  ways
+WHERE
+  ways.tags != ''::hstore AND
+  (
+    (ways.tags?'amenity' AND ways.tags->'amenity' = 'parking') OR
+    (ways.tags?'railway' AND ways.tags->'railway' = 'platform') OR
+    (ways.tags?'aeroway' AND ways.tags->'aeroway' = 'taxiway')
+  )
+UNION ALL
+SELECT
+  id
+FROM
+  nodes
+WHERE
+  tags != ''::hstore AND
+  (
+    (tags?'amenity'  AND tags->'amenity' = 'parking_entrance') OR
+    (tags?'entrance' AND tags->'entrance' IN ('garage', 'emergency'))
+  )
+UNION ALL
+SELECT DISTINCT
+  unnest(nodes) AS id
+FROM
+  highways
+WHERE
+  -- Non-oneway highways are valid input nodes for oneways
+  (NOT is_oneway AND NOT is_roundabout) OR
+  -- Force motorways as valid input nodes
+  -- Raceways are commonly isolated from the main network
+  -- Escape (emergency stop) ways are supposed to be not used / dead-ended
+  -- Footway can lead to anything (attraction entrances, ...) so exclude them even if oneway
+  highway IN ('motorway', 'raceway', 'escape', 'footway') OR
+  -- Construction roads are usually temporary, the connections are likely access=no or similar
+  is_construction
+UNION ALL
+-- Include nodes of ways that cross the extract border. See #1949
+-- Better would be to only include the nodes outside of the extract, but that's expensive
+SELECT
+  unnest(borderways.nodes)
+FROM
+  relation_members AS boundary_members
+  JOIN ways AS boundary_ways ON
+    boundary_members.member_id = boundary_ways.id
+  JOIN highways AS borderways ON
+    ST_Intersects(boundary_ways.linestring, borderways.linestring)
+WHERE
+  boundary_members.member_type = 'W' AND
+  boundary_members.relation_id IN {boundary_ids}
+"""
+
+sql33 = """
+CREATE INDEX idx_allowed_nodes_id ON allowed_end_nodes(id);
+"""
+
+sql34 = """
+-- Make table as short as possible: keep only nodes that are actually nodes of oneways
+CREATE TEMP TABLE allowed_oneway_end_nodes AS
+SELECT DISTINCT
+  allowed_end_nodes.id
+FROM
+  allowed_end_nodes
+  JOIN oneway_highway ON
+    -- Note that we cannot literally only get the first and last node, as i.e. in a looped oneway the 'oneway-entrance/exit node' might be in the middle
+    allowed_end_nodes.id = ANY(oneway_highway.nodes) AND
+    oneway_highway.nid_index = 1 -- no need to compare length(nodes) times
+"""
+
+sql35 = """
+CREATE INDEX idx_allowed_oneway_nodes_id ON allowed_oneway_end_nodes(id);
+DROP TABLE allowed_end_nodes;
+"""
+
+sql36 = """
+-- All end nodes of oneway streets
+CREATE TEMP TABLE oneway_terminal AS
+SELECT DISTINCT ON (nid)
+  oneway.id AS wid,
+  oneway.nodes[oneway.nid_index] AS nid,
+  (SELECT geom FROM nodes WHERE id = oneway.nodes[oneway.nid_index]) AS geom,
```

Comment:
Not sure it is efficient.
Not sure if the planner will do the same or not. But I was more thinking about a join on the result of the first select.

Reply:
> Not sure it is efficient.

The old implementation gave (for japan_chubu, the slowest of the extracts I tested on):
```
['Unique  (cost=2565300.11..2565300.21 rows=21 width=49) (actual time=10072.100..10127.713 rows=28966 loops=1)']
['  ->  Sort  (cost=2565300.11..2565300.16 rows=21 width=49) (actual time=10072.097..10092.663 rows=371818 loops=1)']
['        Sort Key: terminal.id, oneway.id']
['        Sort Method: quicksort  Memory: 41337kB']
['        ->  Merge Left Join  (cost=2564506.66..2565299.65 rows=21 width=49) (actual time=9372.642..9927.462 rows=371818 loops=1)']
['              Merge Cond: (terminal.id = allowed_oneway_end_nodes.id)']
['              Filter: (allowed_oneway_end_nodes.* IS NULL)']
['              Rows Removed by Filter: 458495']
['              ->  Sort  (cost=2548983.05..2548993.68 rows=4254 width=84) (actual time=9284.612..9491.013 rows=830313 loops=1)']
['                    Sort Key: terminal.id']
['                    Sort Method: external merge  Disk: 165400kB']
['                    ->  Nested Loop  (cost=2.86..2548726.64 rows=4254 width=84) (actual time=0.082..8367.067 rows=830313 loops=1)']
['                          ->  Seq Scan on oneway_highway oneway  (cost=0.00..34089.47 rows=425447 width=44) (actual time=0.036..407.462 rows=425447 loops=1)']
['                          ->  Bitmap Heap Scan on nodes terminal  (cost=2.86..5.90 rows=1 width=40) (actual time=0.013..0.017 rows=2 loops=425447)']
['                                Recheck Cond: ((id = (oneway.nodes)[1]) OR (id = (oneway.nodes)[oneway.length]))']
['                                Filter: (array_length(array_positions(oneway.nodes, id), 1) = 1)']
['                                Rows Removed by Filter: 0']
['                                Heap Blocks: exact=742047']
['                                ->  BitmapOr  (cost=2.86..2.86 rows=2 width=0) (actual time=0.009..0.009 rows=0 loops=425447)']
['                                      ->  Bitmap Index Scan on pk_nodes  (cost=0.00..1.43 rows=1 width=0) (actual time=0.004..0.004 rows=1 loops=425447)']
['                                            Index Cond: (id = (oneway.nodes)[1])']
['                                      ->  Bitmap Index Scan on pk_nodes  (cost=0.00..1.43 rows=1 width=0) (actual time=0.003..0.003 rows=1 loops=425447)']
['                                            Index Cond: (id = (oneway.nodes)[oneway.length])']
['              ->  Sort  (cost=15523.61..15909.37 rows=154305 width=40) (actual time=87.495..131.298 rows=575799 loops=1)']
['                    Sort Key: allowed_oneway_end_nodes.id']
['                    Sort Method: quicksort  Memory: 18200kB']
['                    ->  Seq Scan on allowed_oneway_end_nodes  (cost=0.00..2226.05 rows=154305 width=40) (actual time=0.058..37.032 rows=154305 loops=1)']
['Planning Time: 0.222 ms']
['Execution Time: 10136.535 ms']
```

The new implementation gives:
```
['Unique  (cost=41251.58..41251.98 rows=21 width=49) (actual time=5192.359..5202.764 rows=28970 loops=1)']
['  ->  Sort  (cost=41251.58..41251.78 rows=81 width=49) (actual time=5192.356..5195.468 rows=50519 loops=1)']
['        Sort Key: (oneway.nodes[oneway.nid_index]), oneway.id']
['        Sort Method: quicksort  Memory: 5483kB']
['        ->  Nested Loop Left Join  (cost=10.62..41249.01 rows=81 width=49) (actual time=4.009..5158.143 rows=50519 loops=1)']
['              Filter: (allowed_oneway_end_nodes.* IS NULL)']
['              Rows Removed by Filter: 57880']
['              ->  Seq Scan on oneway_highway oneway  (cost=0.00..39407.56 rows=21 width=48) (actual time=0.037..442.005 rows=108399 loops=1)']
['                    Filter: (((nid_index = 1) OR (nid_index = length)) AND (array_length(array_positions(nodes, nodes[nid_index]), 1) = 1))']
['                    Rows Removed by Filter: 317048']
['              ->  Bitmap Heap Scan on allowed_oneway_end_nodes  (cost=10.62..66.14 rows=772 width=40) (actual time=0.006..0.006 rows=1 loops=108399)']
['                    Recheck Cond: ((oneway.nodes)[oneway.nid_index] = id)']
['                    Heap Blocks: exact=57880']
['                    ->  Bitmap Index Scan on idx_allowed_oneway_nodes_id  (cost=0.00..10.42 rows=772 width=0) (actual time=0.003..0.003 rows=1 loops=108399)']
['                          Index Cond: (id = (oneway.nodes)[oneway.nid_index])']
['              SubPlan 1']
['                ->  Index Scan using pk_nodes on nodes  (cost=0.56..3.58 rows=1 width=32) (actual time=0.077..0.077 rows=1 loops=50519)']
['                      Index Cond: (id = (oneway.nodes)[oneway.nid_index])']
['Planning Time: 3.469 ms']
['Execution Time: 5204.168 ms']
```

So it's definitely an significant improvement.

> But I was more thinking about a join on the result of the first select.

I'm unfortunately not sure what you mean?
You want me to change:
```sql
SELECT
  ...
  (SELECT ... ) AS geom
FROM
  ...
```
to:
```sql
SELECT
  t.*,
  nodes.geom
FROM (
   ... current code minus the geom ...
) AS t
JOIN nodes ON
  t.nid == nodes.id
```
?

Reply:
> I'm unfortunately not sure what you mean?
You want me to change:

Yes that my point.

Reply:
Ok, done :)